本日学习任务：<br>
====
一、异常<br>
1、异常概述<br>
异常指不期而至的各种状况，它在程序运行的过程中发生。对于程序的错误以及外部环境能够对用户造成的影响，我们应当及时报告并且以适当的方式来处理这个错误。处理异常，也是为了增强程序的鲁棒性。<br>
鲁棒性:它是在异常和危险情况下系统生存的关键。是指控制系统在一定（结构，大小）的参数摄动下，维持其它某些性能的特性。<br>
异常都是从Throwable类派生出来的，而Throwable类是直接从Object类继承而来。<br>
2、异常分类<br>
(1)异常通常有四类：
```shell
Error：系统内部错误，这类错误由系统进行处理，程序本身无需捕获处理
Exception：可以处理的异常
RuntimeException：可以捕获，也可以不捕获的异常
继承 Exception 的其他类：必须捕获，通常在 API 文档中会说明这些方法抛出哪些异常
```
平时主要关注的异常是 Exception 下的异常，而 Exception 异常下又主要分为两大类异常，一个是派生于 RuntimeExcption 的异常，一个是除了 RuntimeExcption 体系之外的其他异常。<br>
RuntimeExcption 异常(运行时异常)通常有以下几种：
```shell
错误的类型转换
数组访问越界
访问 null 指针
算术异常
```
一般来说，RuntimeException 都是程序员的问题。<br>
非 RuntimeException（受查异常）一般有：
```shell
打开一个不存在的文件
没有找到具有指定名称的类
操作文件异常
```
受查异常是编译器要求必须处理的异常，必须使用 try catch 处理，或者向上抛出，给上层处理。<br>
(2)声明及抛出<br>
throw 抛出异常<br>
当程序运行时数据出现错误或者我们不希望发生的情况出现的话，可以通过抛出异常来处理。<br>
throw 抛出异常语法：throw new 异常名();<br>
实验如下:<br>
![](https://github.com/inspurcloudgroup/rd2/blob/master/%E6%9D%A8%E5%AD%90%E6%B6%B5/0605/img/1.png)<br>
throws 声明异常<br>
throws 用于声明异常，表示该方法可能会抛出的异常。如果声明的异常中包括 checked 异常（受查异常），那么调用者必须处理该异常或者使用 throws 继续向上抛出。<br>
throws 位于方法体前，多个异常使用,分割。<br>
实验如下：<br>
![](https://github.com/inspurcloudgroup/rd2/blob/master/%E6%9D%A8%E5%AD%90%E6%B6%B5/0605/img/2.png)<br>
3、捕获异常<br>
通常抛出异常后，还需要将异常捕获。使用try和catch语句块来捕获异常，有时候还会用到finally。<br>
对于上述三个关键词所构成的语句块，try语句块是必不可少的，catch和finally语句块可以根据情况选择其一或者全选。你可以把可能发生错误或出现问题的语句放到try语句块中，将异常发生后要执行的语句放到catch语句块中，而finally语句块里面放置的语句，不管异常是否发生，它们都会被执行。<br>
捕获异常编程实例<br>
![](https://github.com/inspurcloudgroup/rd2/blob/master/%E6%9D%A8%E5%AD%90%E6%B6%B5/0605/img/3.png)<br>
![](https://github.com/inspurcloudgroup/rd2/blob/master/%E6%9D%A8%E5%AD%90%E6%B6%B5/0605/img/4.png)<br>
4、捕获多个异常<br>
在一段代码中，面对不同的异常，用不同的方式来处理。在这种情况下，可以使用异常匹配，当匹配到对应的异常后，后面的异常将不再进行匹配。<br>
编程实例：<br>
![](https://github.com/inspurcloudgroup/rd2/blob/master/%E6%9D%A8%E5%AD%90%E6%B6%B5/0605/img/5.png)<br>
在处理异常时，并不要求抛出的异常同 catch 所声明的异常完全匹配，子类的对象也可以匹配父类的处理程序。比如异常 A 继承于异常 B，那么在处理多个异常时，一定要将异常 A 放在异常 B 之前捕获，如果将异常 B 放在异常 A 之前，那么将永远匹配到异常 B，异常 A 将永远不可能执行，并且编译器将会报错。如上图异常类A是FileNotFoundException，异常类B是Exception，B是父类，结果编译错误<br>
![](https://github.com/inspurcloudgroup/rd2/blob/master/%E6%9D%A8%E5%AD%90%E6%B6%B5/0605/img/6.png)<br>
5、自定义异常<br>
自定义一个异常类非常简单，只需要让它继承 Exception 或其子类就行。在自定义异常类的时候，建议同时提供无参构造方法和带字符串参数的构造方法，后者可以为你在调试时提供更加详细的信息。<br>
实例如下:<br>
定义异常算术类<br>
![](https://github.com/inspurcloudgroup/rd2/blob/master/%E6%9D%A8%E5%AD%90%E6%B6%B5/0605/img/7.png)<br>
测试用<br>
![](https://github.com/inspurcloudgroup/rd2/blob/master/%E6%9D%A8%E5%AD%90%E6%B6%B5/0605/img/8.png)<br>
编译运行:<br>
![](https://github.com/inspurcloudgroup/rd2/blob/master/%E6%9D%A8%E5%AD%90%E6%B6%B5/0605/img/9.png)<br>
6、异常堆栈<br>
当异常抛出后，我们可以通过异常堆栈追踪程序的运行轨迹，以便我们更好的DEBUG。<br>
实验如下：<br>
![](https://github.com/inspurcloudgroup/rd2/blob/master/%E6%9D%A8%E5%AD%90%E6%B6%B5/0605/img/10.png)<br>
通过上面的异常堆栈轨迹，在对比我们方法的调用过程，可以得出异常信息中首先打印的是距离抛出异常最近的语句，接着是调用该方法的方法，一直到最开始被调用的方法。从下往上看，就可以得出程序运行的轨迹。<br>


